File: src/abstract_factory.java
public abstract class Company {
    public Gpu assembleGpu() {
        Gpu gpu = createGpu();
        gpu.assemble();
        return gpu;
    }

    public abstract Gpu createGpu();
}

public interface Gpu {
    void assemble();
}

public class MsiManufacturer extends Company {
    @Override
    public Gpu createGpu() {
        return new MsiGpu();
    }
}

public class MsiGpu implements Gpu {
    @Override
    public void assemble() {
        // business logic relevant to MSI
    }
}

public class AsusManufacturer extends Company {
    @Override
    public Gpu createGpu() {
        return new AsusGpu();
    }
}

public class AsusGpu implements Gpu {
    @Override
    public void assemble() {
        // business logic relevant to ASUS
    }
}

File: src/creational/abstract_factory.ts
abstract class Company {
  assembleGpu(): Gpu {
    const gpu: Gpu = this.createGpu();
    gpu.assemble();
    return gpu;
  }

  abstract createGpu(): Gpu;
}

interface Gpu {
  assemble(): void;
}

class MsiManufacturer extends Company {
  createGpu(): Gpu {
    return new MsiGpu();
  }
}

class MsiGpu implements Gpu {
  assemble(): void {
    // business logic relevant to MSI
  }
}

class AsusManufacturer extends Company {
  createGpu(): Gpu {
    return new AsusGpu();
  }
}

class AsusGpu implements Gpu {
  assemble(): void {
    // business logic relevant to ASUS
  }
}

const msiManufacturer = new MsiManufacturer();
const msiGpu = msiManufacturer.assembleGpu();

const asusManufacturer = new AsusManufacturer();
const asusGpu = asusManufacturer.assembleGpu();

console.log({ msiGpu });
console.log({ asusGpu });
